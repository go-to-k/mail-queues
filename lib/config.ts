import { StackProps } from "aws-cdk-lib";

export interface Config {
  slackWorkspaceId: string;
  slackChannelId: string;
  senderAddress: string;
}

export interface ConfigStackProps extends StackProps {
  config: Config;
}

export const configStackProps: ConfigStackProps = {
  env: {
    region: "ap-northeast-1",
  },
  config: {
    slackWorkspaceId: "*********",
    slackChannelId: "*********",
    senderAddress: "***@***.***",
  },
};
