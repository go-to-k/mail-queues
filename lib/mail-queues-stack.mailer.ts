import crypto from "crypto";
import fs from "fs";

import { SQSBatchItemFailure, SQSBatchResponse, SQSEvent, SQSHandler } from "aws-lambda";
import * as AWS from "aws-sdk";
import nodemailer from "nodemailer";

import "source-map-support/register";
import { Readable } from "stream";
import {
  MailAttachment,
  MailAttachmentSchema,
  MailParam,
  MailParamSchema,
} from "./types/mailParam";
import { MailRequest, MailRequestSchema } from "./types/mailRequest";

AWS.config.update({ region: process.env.REGION });

const ses = new AWS.SES({
  apiVersion: "2010-12-01",
});

const s3 = new AWS.S3({
  apiVersion: "2010-12-01",
});

const docClient = new AWS.DynamoDB.DocumentClient();

const transporter = nodemailer.createTransport({ SES: ses });

const tableName = process.env.QUEUE_LOCK_TABLE_NAME ?? "";
const attachedFileBucketName = process.env.ATTACHED_FILE_BUCKET_NAME ?? "";
const ttlSecForTable = process.env.TTL_SEC_FOR_TABLE;
const senderAddress = process.env.SENDER_ADDRESS;

export const handler: SQSHandler = async (event: SQSEvent) => {
  const date = new Date();
  const expirationUnixTime: number = Math.floor(date.getTime() / 1000) + Number(ttlSecForTable);

  const batchItemFailureArray: SQSBatchItemFailure[] = [];

  const messageIds: string[] = event.Records.map((record) => {
    return record?.messageId;
  });
  console.log(messageIds);

  for (const record of event.Records) {
    try {
      console.log("messageId: " + record.messageId);

      const mail = (() => {
        try {
          const mailRequest: MailRequest = MailRequestSchema.parse(JSON.parse(record.body));
          return mailRequest;
        } catch (e) {
          console.error(e);
          throw e;
        }
      })();

      const mailParam: MailParam = MailParamSchema.parse({
        from: senderAddress,
        to: mail.toAddress,
        subject: mail.subject,
        text: mail.body,
      });

      const lockMailKey = crypto
        .createHash("sha256")
        .update(mail.mailKey + "-" + mail.toAddress)
        .digest("hex");

      console.log("lockMailKey: " + lockMailKey);

      const isLockSucceeded = await lockTable(lockMailKey, expirationUnixTime).catch((e) => {
        console.error(e);
        throw e;
      });

      if (!isLockSucceeded) {
        continue;
      }

      try {
        if (mail?.attachedFileKeys && mail?.attachedFileKeys.length !== 0) {
          makeDiskDir(mail.mailKey);

          const attachmentsPromises = mail?.attachedFileKeys.map((attachedFileKey) => {
            return getMailAttachmentFromS3(mail.mailKey, attachedFileKey);
          });

          mailParam.attachments = await Promise.allSettled(attachmentsPromises).then((results) =>
            results.map((result) => {
              if (result.status === "rejected") {
                throw new Error(result.reason);
              }
              return MailAttachmentSchema.parse(result.value);
            }),
          );
        }

        await transporter.sendMail(mailParam);

        console.log("Sent mail.");
      } catch (e) {
        console.error(e);
        await unlockTable(lockMailKey);
        throw e;
      }
    } catch (e) {
      const batchItemFailure: SQSBatchItemFailure = {
        itemIdentifier: record.messageId,
      };

      batchItemFailureArray.push(batchItemFailure);
    }
  }

  const sqsBatchResponse: SQSBatchResponse = {
    batchItemFailures: batchItemFailureArray,
  };

  console.log(sqsBatchResponse);

  return sqsBatchResponse;
};

const lockTable = async (lockMailKey: string, expirationUnixTime: number): Promise<boolean> => {
  if (!lockMailKey) {
    throw new Error("Empty lockMailKey.");
  }

  const params: AWS.DynamoDB.DocumentClient.PutItemInput = {
    TableName: tableName,
    Item: {
      LockMailKey: lockMailKey,
      ExpirationUnixTime: expirationUnixTime,
    },
    ConditionExpression: "attribute_not_exists(#hash)",
    ExpressionAttributeNames: {
      "#hash": "LockMailKey",
    },
  };

  return await docClient
    .put(params)
    .promise()
    .then(() => {
      return true;
    })
    .catch((e) => {
      if (e?.code === "ConditionalCheckFailedException") {
        console.log("Lock Already Exists.");
        return false;
      } else {
        console.error("Other DynamoDB Lock Error.");
        throw e;
      }
    });
};

const unlockTable = async (lockMailKey: string): Promise<boolean> => {
  const params: AWS.DynamoDB.DocumentClient.DeleteItemInput = {
    TableName: tableName,
    Key: {
      LockMailKey: lockMailKey,
    },
    ConditionExpression: "attribute_exists(#hash)",
    ExpressionAttributeNames: {
      "#hash": "LockMailKey",
    },
  };

  return await docClient
    .delete(params)
    .promise()
    .then(() => {
      console.log("Unlock Success.");
      return true;
    })
    .catch((e) => {
      console.error("Other DynamoDB Unlock Error.");
      return false;
    });
};

const makeDiskDir = (diskDirName: string) => {
  const diskDirPath = `/tmp/${diskDirName}`;

  try {
    if (!fs.existsSync(diskDirPath)) {
      fs.mkdirSync(diskDirPath);
    }
  } catch (e) {
    console.error(e);
    throw e;
  }
};

const getMailAttachmentFromS3 = async (
  mailKey: string,
  attachedFileKey: string,
): Promise<MailAttachment> => {
  const s3Key = `${mailKey}/${attachedFileKey}`;
  const diskFilePath = `/tmp/${s3Key}`;

  const attachment: MailAttachment = {
    filename: attachedFileKey,
    path: diskFilePath,
  };

  try {
    const { Body } = await s3
      .getObject({
        Bucket: attachedFileBucketName,
        Key: s3Key,
      })
      .promise();

    await fs.promises.writeFile(diskFilePath, Body as Readable);

    console.log("Done:", s3Key);
  } catch (e) {
    console.error(e + ":" + s3Key);
    throw e;
  }

  return attachment;
};
