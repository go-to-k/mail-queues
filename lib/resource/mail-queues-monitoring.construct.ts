import { Duration } from "aws-cdk-lib";
import { SlackChannelConfiguration } from "aws-cdk-lib/aws-chatbot";
import { Alarm, ComparisonOperator, TreatMissingData } from "aws-cdk-lib/aws-cloudwatch";
import { SnsAction } from "aws-cdk-lib/aws-cloudwatch-actions";
import { Topic } from "aws-cdk-lib/aws-sns";
import { DeadLetterQueue, Queue, QueueEncryption } from "aws-cdk-lib/aws-sqs";
import { Construct } from "constructs";
import { StackInput } from "../types/stack-input";

export type MailQueuesMonitoringProps = Pick<StackInput, "slackWorkspaceId" | "slackChannelId"> & {
  slackChannelConfigurationName: string;
};

export class MailQueuesMonitoring extends Construct {
  public readonly deadLetterQueue: DeadLetterQueue;

  constructor(scope: Construct, id: string, props: MailQueuesMonitoringProps) {
    super(scope, id);

    const dlq = new Queue(this, "DeadLetterQueue", {
      retentionPeriod: Duration.seconds(1209600),
      encryption: QueueEncryption.KMS_MANAGED,
    });

    this.deadLetterQueue = {
      queue: dlq,
      maxReceiveCount: 5,
    };

    const errorTopic = new Topic(this, "ErrorTopic", {
      displayName: "ErrorTopic",
      topicName: "ErrorTopic",
    });

    new SlackChannelConfiguration(this, "ErrorSlackChatbotConfiguration", {
      slackChannelConfigurationName: props.slackChannelConfigurationName,
      slackWorkspaceId: props.slackWorkspaceId,
      slackChannelId: props.slackChannelId,
      notificationTopics: [errorTopic],
    });

    const dlqSizeAlarm = new Alarm(this, "DeadLetterQueueSizeAlarm", {
      metric: dlq.metricApproximateNumberOfMessagesVisible().with({
        period: Duration.seconds(60),
        statistic: "Sum",
      }),
      threshold: 1,
      evaluationPeriods: 1,
      datapointsToAlarm: 1,
      comparisonOperator: ComparisonOperator.GREATER_THAN_OR_EQUAL_TO_THRESHOLD,
      treatMissingData: TreatMissingData.NOT_BREACHING,
    });

    dlqSizeAlarm.addAlarmAction(new SnsAction(errorTopic));
  }
}
