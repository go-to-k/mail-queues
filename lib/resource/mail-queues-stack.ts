import { CfnOutput, Duration, Stack } from "aws-cdk-lib";
import { Construct } from "constructs";
import { NodejsFunction } from "aws-cdk-lib/aws-lambda-nodejs";
import { Queue, QueueEncryption, DeadLetterQueue } from "aws-cdk-lib/aws-sqs";
import { AttributeType, BillingMode, Table, TableEncryption } from "aws-cdk-lib/aws-dynamodb";
import { SqsEventSource } from "aws-cdk-lib/aws-lambda-event-sources";
import { Bucket, BucketEncryption } from "aws-cdk-lib/aws-s3";
import { Topic } from "aws-cdk-lib/aws-sns";
import { SlackChannelConfiguration } from "aws-cdk-lib/aws-chatbot";
import { Alarm, ComparisonOperator, TreatMissingData } from "aws-cdk-lib/aws-cloudwatch";
import { SnsAction } from "aws-cdk-lib/aws-cloudwatch-actions";
import { Effect, PolicyStatement } from "aws-cdk-lib/aws-iam";
import { StackValidator } from "../validator/stack-validator";
import { ConfigStackProps } from "../config";

export class MailQueuesStack extends Stack {
  private slackWorkspaceId: string;
  private slackChannelId: string;
  private senderAddress: string;

  constructor(scope: Construct, id: string, props: ConfigStackProps) {
    super(scope, id, props);

    this.init(props);
    this.create();
  }

  private init(props: ConfigStackProps): void {
    this.slackWorkspaceId = props.config.slackWorkspaceId;
    this.slackChannelId = props.config.slackChannelId;
    this.senderAddress = props.config.senderAddress;

    const stackValidator = new StackValidator(
      this.slackWorkspaceId,
      this.slackChannelId,
      this.senderAddress,
    );
    this.node.addValidation(stackValidator);
  }

  private create() {
    const attachedFileBucket = new Bucket(this, "AttachedFileBucket", {
      encryption: BucketEncryption.KMS_MANAGED,
    });

    const table = new Table(this, "QueueLockTable", {
      partitionKey: {
        name: "LockMailKey",
        type: AttributeType.STRING,
      },
      billingMode: BillingMode.PAY_PER_REQUEST,
      pointInTimeRecovery: true,
      encryption: TableEncryption.DEFAULT,
      timeToLiveAttribute: "ExpirationUnixTime",
    });

    const mailerHandler = new NodejsFunction(this, "mailer", {
      environment: {
        REGION: this.region,
        QUEUE_LOCK_TABLE_NAME: table.tableName,
        ATTACHED_FILE_BUCKET_NAME: attachedFileBucket.bucketName,
        TTL_SEC_FOR_TABLE: "345600", // seconds
        SENDER_ADDRESS: this.senderAddress,
      },
      bundling: {
        forceDockerBundling: false,
      },
    });

    attachedFileBucket.grantRead(mailerHandler);
    table.grantReadWriteData(mailerHandler);
    mailerHandler.addToRolePolicy(
      new PolicyStatement({
        actions: ["ses:SendEmail", "SES:SendRawEmail"],
        resources: ["*"],
        effect: Effect.ALLOW,
      }),
    );

    const dlq = new Queue(this, "DeadLetterQueue", {
      retentionPeriod: Duration.seconds(1209600),
      encryption: QueueEncryption.KMS_MANAGED,
    });

    const deadLetterQueue: DeadLetterQueue = {
      queue: dlq,
      maxReceiveCount: 5,
    };

    const queue = new Queue(this, "MailQueue", {
      visibilityTimeout: Duration.seconds(30),
      receiveMessageWaitTime: Duration.seconds(10),
      encryption: QueueEncryption.KMS_MANAGED,
      deadLetterQueue: deadLetterQueue,
    });

    const eventSource = new SqsEventSource(queue, {
      batchSize: 5,
      maxBatchingWindow: Duration.seconds(10),
      reportBatchItemFailures: true,
    });

    mailerHandler.addEventSource(eventSource);

    const errorTopic = new Topic(this, "ErrorTopic", {
      displayName: "ErrorTopic",
      topicName: "ErrorTopic",
    });

    new SlackChannelConfiguration(this, "ErrorSlackChatbotConfiguration", {
      slackChannelConfigurationName: this.stackName,
      slackWorkspaceId: this.slackWorkspaceId,
      slackChannelId: this.slackChannelId,
      notificationTopics: [errorTopic],
    });

    const dlqSizeAlarm = new Alarm(this, "DeadLetterQueueSizeAlarm", {
      metric: dlq.metricApproximateNumberOfMessagesVisible().with({
        period: Duration.seconds(60),
        statistic: "Sum",
      }),
      threshold: 1,
      evaluationPeriods: 1,
      datapointsToAlarm: 1,
      comparisonOperator: ComparisonOperator.GREATER_THAN_OR_EQUAL_TO_THRESHOLD,
      treatMissingData: TreatMissingData.NOT_BREACHING,
    });

    dlqSizeAlarm.addAlarmAction(new SnsAction(errorTopic));

    new CfnOutput(this, "MailQueueUrl", {
      value: queue.queueUrl,
      exportName: `${this.stackName}-MailQueueUrl`,
    });
  }
}
