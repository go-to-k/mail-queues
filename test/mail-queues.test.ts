import { App, assertions } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { ConfigStackProps } from "../lib/config";
import { MailQueuesStack } from "../lib/resource/mail-queues-stack";

const getTemplate = (input: {
  slackWorkspaceId: string;
  slackChannelId: string;
  senderAddress: string;
  region: string;
}): assertions.Template => {
  const { slackWorkspaceId, slackChannelId, senderAddress, region } = input;

  const testSlackWorkspaceId = slackWorkspaceId;
  const testSlackChannelId = slackChannelId;
  const testSenderAddress = senderAddress;
  const testRegion = region;

  const testConfigStackProps: ConfigStackProps = {
    env: {
      region: testRegion,
    },
    config: {
      slackWorkspaceId: testSlackWorkspaceId,
      slackChannelId: testSlackChannelId,
      senderAddress: testSenderAddress,
    },
  };

  const app = new App();
  const stack = new MailQueuesStack(app, "MailQueuesStack", testConfigStackProps);
  return Template.fromStack(stack);
};

describe("Snapshot Tests", () => {
  const region = "ap-northeast-1";
  const input = {
    slackWorkspaceId: "slackWorkspaceId",
    slackChannelId: "slackChannelId",
    senderAddress: "senderAddress@test.test",
    region,
  };
  const template = getTemplate(input);

  test("Snapshot test", () => {
    expect(template.toJSON()).toMatchSnapshot();
  });
});

describe("Fine-grained Assertions Tests", () => {
  const region = "ap-northeast-1";
  const input = {
    slackWorkspaceId: "slackWorkspaceId",
    slackChannelId: "slackChannelId",
    senderAddress: "senderAddress@test.test",
    region,
  };
  const template = getTemplate(input);

  test("Bucket created", () => {
    template.resourceCountIs("AWS::S3::Bucket", 1);
  });

  test("DynamoDB created", () => {
    template.resourceCountIs("AWS::DynamoDB::Table", 1);
  });

  test("IAM Role created", () => {
    template.resourceCountIs("AWS::IAM::Role", 3); // +1 for "autoDeleteObjects: true" in the Bucket
  });

  test("IAM Policy created", () => {
    template.resourceCountIs("AWS::IAM::Policy", 1);
  });

  test("Lambda created", () => {
    template.resourceCountIs("AWS::Lambda::Function", 2); // +1 for "autoDeleteObjects: true" in the Bucket
  });

  test("EventSourceMapping created", () => {
    template.resourceCountIs("AWS::Lambda::EventSourceMapping", 1);
  });

  test("SQS created", () => {
    template.resourceCountIs("AWS::SQS::Queue", 2);
  });

  test("Topic created", () => {
    template.resourceCountIs("AWS::SNS::Topic", 1);
  });

  test("Chatbot created", () => {
    template.resourceCountIs("AWS::Chatbot::SlackChannelConfiguration", 1);
  });

  test("Alarm created", () => {
    template.resourceCountIs("AWS::CloudWatch::Alarm", 1);
  });
});

describe("Validation Tests", () => {
  const region = "ap-northeast-1";

  test("slackWorkspaceId is empty", () => {
    const input = {
      slackWorkspaceId: "",
      slackChannelId: "slackChannelId",
      senderAddress: "senderAddress@test.test",
      region,
    };

    expect(() => {
      getTemplate(input);
    }).toThrow(
      /{"issues":\[{"code":"too_small","minimum":1,"type":"string","inclusive":true,"message":"String must contain at least 1 character\(s\)","path":\["slackWorkspaceId"\]}\],"name":"ZodError"}/,
    );
  });

  test("slackChannelId is empty", () => {
    const input = {
      slackWorkspaceId: "slackWorkspaceId",
      slackChannelId: "",
      senderAddress: "senderAddress@test.test",
      region,
    };

    expect(() => {
      getTemplate(input);
    }).toThrow(
      /{"issues":\[{"code":"too_small","minimum":1,"type":"string","inclusive":true,"message":"String must contain at least 1 character\(s\)","path":\["slackChannelId"\]}\],"name":"ZodError"}/,
    );
  });

  test("senderAddress is empty", () => {
    const input = {
      slackWorkspaceId: "slackWorkspaceId",
      slackChannelId: "slackChannelId",
      senderAddress: "",
      region,
    };

    expect(() => {
      getTemplate(input);
    }).toThrow(
      /{"issues":\[{"validation":"email","code":"invalid_string","message":"Invalid email","path":\["senderAddress"\]},{"code":"too_small","minimum":1,"type":"string","inclusive":true,"message":"String must contain at least 1 character\(s\)","path":\["senderAddress"\]}\],"name":"ZodError"}/,
    );
  });

  test("senderAddress is invalid", () => {
    const input = {
      slackWorkspaceId: "slackWorkspaceId",
      slackChannelId: "slackChannelId",
      senderAddress: "senderAddress.test.test",
      region,
    };

    expect(() => {
      getTemplate(input);
    }).toThrow(
      /{"issues":\[{"validation":"email","code":"invalid_string","message":"Invalid email","path":\["senderAddress"\]}\],"name":"ZodError"}/,
    );
  });
});
