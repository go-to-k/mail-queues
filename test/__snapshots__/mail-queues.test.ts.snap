// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Tests Snapshot test 1`] = `
Object {
  "Outputs": Object {
    "MailQueueUrl": Object {
      "Export": Object {
        "Name": "MailQueuesStack-MailQueueUrl",
      },
      "Value": Object {
        "Ref": "MailQueue4F838B3A",
      },
    },
  },
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "AttachedFileBucketAutoDeleteObjectsCustomResource907B3BD5": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "AttachedFileBucketPolicyCF04CCE1",
      ],
      "Properties": Object {
        "BucketName": Object {
          "Ref": "AttachedFileBucketC6C49BBC",
        },
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "CustomS3AutoDeleteObjectsCustomResourceProviderHandler9D90184F",
            "Arn",
          ],
        },
      },
      "Type": "Custom::S3AutoDeleteObjects",
      "UpdateReplacePolicy": "Delete",
    },
    "AttachedFileBucketC6C49BBC": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "BucketEncryption": Object {
          "ServerSideEncryptionConfiguration": Array [
            Object {
              "ServerSideEncryptionByDefault": Object {
                "SSEAlgorithm": "AES256",
              },
            },
          ],
        },
        "LifecycleConfiguration": Object {
          "Rules": Array [
            Object {
              "ExpirationInDays": 400,
              "Id": "delete-lifecycle",
              "Status": "Enabled",
            },
          ],
        },
        "Tags": Array [
          Object {
            "Key": "aws-cdk:auto-delete-objects",
            "Value": "true",
          },
        ],
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Delete",
    },
    "AttachedFileBucketPolicyCF04CCE1": Object {
      "Properties": Object {
        "Bucket": Object {
          "Ref": "AttachedFileBucketC6C49BBC",
        },
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
              ],
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::GetAtt": Array [
                    "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
                    "Arn",
                  ],
                },
              },
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "AttachedFileBucketC6C49BBC",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "AttachedFileBucketC6C49BBC",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "CustomS3AutoDeleteObjectsCustomResourceProviderHandler9D90184F": Object {
      "DependsOn": Array [
        "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-ap-northeast-1",
          },
          "S3Key": "HASH-REPLACED.zip",
        },
        "Description": Object {
          "Fn::Join": Array [
            "",
            Array [
              "Lambda function for auto-deleting objects in ",
              Object {
                "Ref": "AttachedFileBucketC6C49BBC",
              },
              " S3 bucket.",
            ],
          ],
        },
        "Handler": "__entrypoint__.handler",
        "MemorySize": 128,
        "Role": Object {
          "Fn::GetAtt": Array [
            "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Sub": "arn:\${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "MailQueue4F838B3A": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "KmsMasterKeyId": "alias/aws/sqs",
        "ReceiveMessageWaitTimeSeconds": 10,
        "RedrivePolicy": Object {
          "deadLetterTargetArn": Object {
            "Fn::GetAtt": Array [
              "MonitoringDeadLetterQueue5C1A4660",
              "Arn",
            ],
          },
          "maxReceiveCount": 5,
        },
        "VisibilityTimeout": 30,
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "MonitoringDeadLetterQueue5C1A4660": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "KmsMasterKeyId": "alias/aws/sqs",
        "MessageRetentionPeriod": 1209600,
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "MonitoringDeadLetterQueueSizeAlarmB6C2646B": Object {
      "Properties": Object {
        "AlarmActions": Array [
          Object {
            "Ref": "MonitoringErrorTopic16B8889E",
          },
        ],
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "DatapointsToAlarm": 1,
        "Dimensions": Array [
          Object {
            "Name": "QueueName",
            "Value": Object {
              "Fn::GetAtt": Array [
                "MonitoringDeadLetterQueue5C1A4660",
                "QueueName",
              ],
            },
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "ApproximateNumberOfMessagesVisible",
        "Namespace": "AWS/SQS",
        "Period": 60,
        "Statistic": "Sum",
        "Threshold": 1,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "MonitoringErrorSlackChatbotConfiguration6AB0E891": Object {
      "Properties": Object {
        "ConfigurationName": "MailQueuesStack",
        "IamRoleArn": Object {
          "Fn::GetAtt": Array [
            "MonitoringErrorSlackChatbotConfigurationConfigurationRoleEC8FD28D",
            "Arn",
          ],
        },
        "SlackChannelId": "slackChannelId",
        "SlackWorkspaceId": "slackWorkspaceId",
        "SnsTopicArns": Array [
          Object {
            "Ref": "MonitoringErrorTopic16B8889E",
          },
        ],
      },
      "Type": "AWS::Chatbot::SlackChannelConfiguration",
    },
    "MonitoringErrorSlackChatbotConfigurationConfigurationRoleEC8FD28D": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "chatbot.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "MonitoringErrorTopic16B8889E": Object {
      "Properties": Object {
        "DisplayName": "ErrorTopic",
        "TopicName": "ErrorTopic",
      },
      "Type": "AWS::SNS::Topic",
    },
    "QueueLockTable78CED995": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "AttributeDefinitions": Array [
          Object {
            "AttributeName": "LockMailKey",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "KeySchema": Array [
          Object {
            "AttributeName": "LockMailKey",
            "KeyType": "HASH",
          },
        ],
        "PointInTimeRecoverySpecification": Object {
          "PointInTimeRecoveryEnabled": true,
        },
        "SSESpecification": Object {
          "SSEEnabled": false,
        },
        "TimeToLiveSpecification": Object {
          "AttributeName": "ExpirationUnixTime",
          "Enabled": true,
        },
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Delete",
    },
    "mailerE5F4EB7C": Object {
      "DependsOn": Array [
        "mailerServiceRoleDefaultPolicyDEC8276C",
        "mailerServiceRoleF01D1C35",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-ap-northeast-1",
          },
          "S3Key": "HASH-REPLACED.zip",
        },
        "Environment": Object {
          "Variables": Object {
            "ATTACHED_FILE_BUCKET_NAME": Object {
              "Ref": "AttachedFileBucketC6C49BBC",
            },
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "QUEUE_LOCK_TABLE_NAME": Object {
              "Ref": "QueueLockTable78CED995",
            },
            "REGION": "ap-northeast-1",
            "SENDER_ADDRESS": "senderAddress@test.test",
            "TTL_SEC_FOR_TABLE": "345600",
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "mailerServiceRoleF01D1C35",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "mailerServiceRoleDefaultPolicyDEC8276C": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "AttachedFileBucketC6C49BBC",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "AttachedFileBucketC6C49BBC",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": Array [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:ConditionCheckItem",
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
                "dynamodb:DescribeTable",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "QueueLockTable78CED995",
                    "Arn",
                  ],
                },
                Object {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
            Object {
              "Action": Array [
                "ses:SendEmail",
                "SES:SendRawEmail",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "sqs:ReceiveMessage",
                "sqs:ChangeMessageVisibility",
                "sqs:GetQueueUrl",
                "sqs:DeleteMessage",
                "sqs:GetQueueAttributes",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "MailQueue4F838B3A",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "mailerServiceRoleDefaultPolicyDEC8276C",
        "Roles": Array [
          Object {
            "Ref": "mailerServiceRoleF01D1C35",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "mailerServiceRoleF01D1C35": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "mailerSqsEventSourceMailQueuesStackMailQueue9C289D8422611485": Object {
      "Properties": Object {
        "BatchSize": 5,
        "EventSourceArn": Object {
          "Fn::GetAtt": Array [
            "MailQueue4F838B3A",
            "Arn",
          ],
        },
        "FunctionName": Object {
          "Ref": "mailerE5F4EB7C",
        },
        "FunctionResponseTypes": Array [
          "ReportBatchItemFailures",
        ],
        "MaximumBatchingWindowInSeconds": 10,
      },
      "Type": "AWS::Lambda::EventSourceMapping",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
